river/Cursor.zig

```diff
@@ -40,6 +40,108 @@ const View = @import("View.zig");
 
+pub const ResizeDirection = enum {
+    top,
+    right,
+    bottom,
+    left,
+    topleft,
+    topright,
+    botleft,
+    botright,
+
+    /// Returns the resize direction that makes the most sense based on the
+    /// position of the cursor on the current surface.
+    pub fn fromCursorPosition(cursor: *Self, view: *View) ResizeDirection {
+        const view_box = &view.current.box;
+        const output_box = server.root.output_layout.getBox(view.output.wlr_output).?;
+        const pointer_surface_x = (@floatToInt(i32, cursor.wlr_cursor.x) - output_box.x) - view_box.x;
+        const pointer_surface_y = (@floatToInt(i32, cursor.wlr_cursor.y) - output_box.y) - view_box.y;
+
+        const Position = enum {
+            start, // top or left
+            center,
+            end, // bottom or right
+
+            /// Arbitrary minimum size of resize "handle" area. Must be smaller
+            /// than View.min_size / 2, as otherwise an integer underflow can
+            /// occur.
+            const min_size = 20;
+
+            fn fromCoords(pos: i32, size: u32) @This() {
+                const handle: u32 = std.math.max(min_size, size / 5);
+                if (pos < handle) return .start;
+                if (pos > size - handle) return .end;
+                return .center;
+            }
+        };
+
+        const horizontal = Position.fromCoords(pointer_surface_x, view_box.width);
+        const vertical = Position.fromCoords(pointer_surface_y, view_box.height);
+
+        // The compiler complains when returning on outer switch...
+        switch (horizontal) {
+            .start => return switch (vertical) {
+                .start => .topleft,
+                .center => .left,
+                .end => .botleft,
+            },
+            .center => return switch (vertical) {
+                .start => .top,
+                .center => .botright, // the center should not be a dead-zone
+                .end => .bottom,
+            },
+            .end => return switch (vertical) {
+                .start => .topright,
+                .center => .right,
+                .end => .botright,
+            },
+        }
+    }
+
+    /// Tries to convert wlr.Edges to ResizeDirection.
+    pub fn fromWlrEdges(edges: wlr.Edges) !ResizeDirection {
+        // Only two edges may be active at max and they must not be opposing.
+        if (edges.top and edges.bottom) return error.InvalidEdges;
+        if (edges.right and edges.left) return error.InvalidEdges;
+
+        if (edges.top) {
+            if (edges.right) return .topright;
+            if (edges.left) return .topleft;
+            return .top;
+        }
+
+        if (edges.bottom) {
+            if (edges.right) return .botright;
+            if (edges.left) return .botleft;
+            return .bottom;
+        }
+
+        if (edges.right) return .right;
+
+        if (edges.left) return .left;
+
+        return error.InvalidEdges;
```

Comment:
I'm pretty sure this needs to handle the case in which no edges are set as well. `none` is a valid value of the `resize_edge` enum in the protocol. In this case let's default to bottom right.
```suggestion
        return .botright;
```